### ISSUE TITLE
preencoding of text to memmapped embeddings

### ISSUE THREAD

**Issue by TheFusion21** [lucidrains/imagen-pytorch#266]
- When should this happen?
- Should this be with file?
- per batch or the whole dataset?

**Comment by lucidrains** [comment]
@TheFusion21 ideally always is my opinion! i really like this package https://github.com/hristo-vrigazov/mmap.ninja

**Comment by TheFusion21** [comment]
do this when adding the dataset to the trainer?
and use the batchsize the same as for the dataset?

**Comment by TheFusion21** [comment]
how to handle custom dataset and collate_fn

**Comment by TheFusion21** [comment]
so I could wrap the dataset but dont know what data I get and the same goes for the collate.
I know what data I can get but it can be shuffled so I have no way to index it.
I could wrap the dataset easily if I get user provided labels for the text which i could map and return with the rest of the data from this dataset

**Comment by jacobwjs** [comment]
@TheFusion21 Sorry for pinging across multiple threads on this. To follow up on your questions, this would be a precomputed approach before any training run. There’s no real benefit to doing it during training/instantiating the dataset especially when training a something large (> 500M examples). You’ll be waiting for quite sometime before training kicks off, and disk space will be a big issue.

So should do it before training to save valuable VRAM and free up larger batch size for computing embeddings.

So, precompute against entire text portion of the (image, text) dataset (e.g. COYO-700M, LAION, etc) and store in a way that can be matched up against shards and individual examples from what is most likely a `webdataset` containing your favorite image captioned dataset.

Suppose you could stuff this all in a `webdataset`, but then you’re tightly coupled to T5 model but batching becomes easier.

**Comment by TheFusion21** [comment]
Hey @jacobwjs,

haven't had time at all. Busy with work ^^.
Im not to familiar with webdataset, but I would just add a CLI command to precompute a given Dataset with selected T5 model.
I'm currently doing it during training for my data at least on CPU as I have plenty of headroom to spare.
It is fast enough for me on one GPU. I guess if you train fast enough that will bottleneck pretty fast.

**Comment by jacobwjs** [comment]
Thanks for the reply. 

Sounds like we’re working with very different datasets and model sizes. I’ll forge ahead then if nothings been done :)
